### ISSUE TITLE
Refactor the way Android Plugins are included

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#29328]
Today, we include plugins directly as subprojects. This causes problems when there are dependency conflicts, and tools like Jetifier that should be able to help us resolve them aren't set up to run on subprojects (they assume the project will more or less internally consistent).  In particular, this has caused problems for Android X (#23586  https://github.com/flutter/flutter/issues/27106).  It's also meaningful for Add2App use cases, where we have customers who want to distribute Flutter in a more modular fashion (e.g. as AARs rather than as source code).  The main concern right now is solving the Android X headaches though. The following dependencies in pubspec.yaml would cause an android project to fail to build because of such conflicts:

```yaml
dependencies:
  shared_preferences: any
  flutter:
    sdk: flutter
  camera: 0.2.6
  firebase_auth: any
```

A quick POC resolves a lot of these problems:

- Include plugins as `includeBuild` in settings.gradle rather than as a subproject.
- Specify dependencies on plugins in the `dependencies` section of `app/build.gradle`.
- For each plugin, make sure it has  all of its dependencies specified correctly in its `build.gradle` (including a dependency on flutter.jar).

For example:

Change android/settings.gradle's plugins.each to:
```groovy
plugins.each { name, path ->
    includeBuild "$path/android"
}
```

Remove the subprojects from android/build.gradle and align the gradle plugin with other plugins' requirements:
```groovy
...
        classpath 'com.android.tools.build:gradle:3.3.1'
...
rootProject.buildDir = '../build'
// subprojects {
//     project.buildDir = "${rootProject.buildDir}/${project.name}"
// }
// subprojects {
//     project.evaluationDependsOn(':app')
// }
...
```

Add the right dependencies to android/app/build.gradle:
```groovy
dependencies {
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'com.android.support.test:runner:1.0.2'
    androidTestImplementation 'com.android.support.test.espresso:espresso-core:3.0.2'
    implementation 'io.flutter.plugins.sharedpreferences:shared_preferences:1.0.0'
    implementation 'io.flutter.plugins.firebase.core:firebase_core:1.0.0'
    implementation 'io.flutter.plugins.firebaseauth:firebase_auth:1.0.0'
    implementation 'io.flutter.plugins.camera:camera:1.0.0'
}
```

This also requires changes in each plugin's build.gradle, at a minimum:
```groovy
dependencies {
    api files('/Path/to/flutter/bin/cache/artifacts/engine/android-arm/flutter.jar')
}
```

The `includeBuild` method will get Gradle to build all these projects with one command, but is not ideal because it requires that all projects use the same gradle plugin version.

To remedy this, we should teach the flutter tool and/or flutter.gradle how to build Android plugins directly. This may end up including updating the plugin's gradle configuration so that it more correctly specifies its dependencies, including a dependency on flutter.jar from the framework/engine.  We need to figure out if we can do that purely through the DSL or whether we'd actually have to generate a new build.gradle file for these purposes. 

/cc @mklim @amirh @matthew-carroll @csells

**Comment by dnfield** [comment]
Depending on what the solution ends up looking like in the end, we may want to distribute some of this work independently from Flutter tool - perhaps as an independent gradle plugin or as Dart package - so that we could retroactively apply this to stable channel without requiring an upgrade.

**Comment by mklim** [comment]
I've had to tweak these steps slightly to get the proof of concept working on my machine.

- Copied the `android/local.properties` from my app to the android dir of all the different plugin repos to get around a missing sdk error. I do have `$ANDROID_HOME` set in my environment correctly so this shouldn't theoretically be necessary, probably some strange things going on there.
- Added the flutter jar dependency as a `compileOnly` instead of `api` to avoid a duplicate classpath conflict.

**Comment by mklim** [comment]
I put up a very rough WIP of building the plugins as AARs and then including them at runtime at flutter/flutter#29460. It's not ready to merge yet though, mostly because it's still blocked on two larger problems:

1. It requires apps to manually remove the `subprojects` declarations from `android/build.gradle` and the `plugins.each` loop from `android/settings.gradle` like mentioned in the first comment of this issue. I think arguably this shouldn't require manual intervention to work at all, but if it does we need to gate it behind a check for the changes and use the current logic still otherwise.
2. Transitive gradle dependencies within our Flutter plugins cause crashes at runtime, since the inner deps aren't included. As it is the patch requires apps to manually include all jars depended on by their plugins at runtime

**Comment by rmtmckenzie** [comment]
Since this is being worked on now, something to consider... some of the APIs from the compat/androidx libraries depend on FragmentActivity being passed to them rather than a Context or Activity. In any plugins in the current system that basically means trying to cast the activity that is passed into the plugin to FragmentActivity and failing if the activity isn't a compat/androidx library.

A good example of this is the androidx biometric library's BiometricPrompt which takes a FragmentActivity.

Also see issue #29819 which describes a problem with FlutterFragmentActivity when using the AndroidX libraries - I don't know if that's something that would be affected by this work or not.

**Comment by dnfield** [comment]
@rmtmckenzie - @matthew-carroll is working on a larger project to make that possible AFAIK

**Comment by matthew-carroll** [comment]
FYI @dnfield @mklim I've got a thread going on over in #29819 that will probably resolve @rmtmckenzie's comment, but it's unclear what the solution is.  We're facing an AndroidX vs support lib problem.  Have either of your efforts treated that issue at-large?

**Comment by mklim** [comment]
@matthew-carroll that's sort of what this issue is tracking, but it's impossible to completely solve this issue at large since the two sets of libraries are fundamentally incompatible. Like @dnfield mentioned earlier in this issue, Jetifier should be able to solve AndroidX incompatibilities when an app is including a `.aar` as a dependency by replacing the classpaths in the compiled dependency to match the new libraries. But our current build drags in the different Flutter plugin dependencies as subprojects, so any incompatibilities break down at compile time instead. This work is splitting out plugins so that they're external Java dependencies that can be Jetified. Even with this change though trying to actually include both libraries in the same Flutter app source code simultaneously would still fail. Responded on that issue with some follow up questions, I think it should be handled by our build structure already if I'm understanding it right.

**Comment by dnfield** [comment]
It's looking like our real problem is this:

```
subprojects {
    project.evaluationDependsOn(':app')
}
```

Removing that lets Jetifier do its thing properly.

I think our immediate fix is to just make that work.  We can revisit the possibility of just building and publishing AARs from Flutter plugins separately.

**Comment by xster** [comment]
@blasten also to keep on your mind as you perform refactors

**Comment by xster** [comment]
This might be accounted for above already, but this also necessarily implies having a solution for https://github.com/flutter/flutter/issues/22102. Since our include_flutter.groovy'ed project can't bring in resources currently (like our plugins currently can as source or, in the future, as AARs)

**Comment by xster** [comment]
Assigning to @blasten since you're actively working on this.

**Comment by mcxinyu** [comment]
> 
> ```
> subprojects {
>    // project.evaluationDependsOn(':app')
> }
> ```


Error: Package io.flutter.plugin.common does not exist

**Comment by xster** [comment]
@blasten we can close this issue now right? (with some instructions on how the fallbacks work and how the flag works)

**Comment by blasten** [comment]
tl;dr: we still include plugins as subprojects and attempt to include them as AAR (Maven dependencies) when the build failed likely due to an AndroidX issue. This is a hack to get Jetifier to work with Flutter plugins.

The ecosystem team mentioned that they will still prefer to change the default, so plugins are always included as Maven dependencies.  However, this isn't trivial and involves non-trivial risks:

e.g. changes to the serving flow (pub). Tooling on the authoring side, so it's easy to verify that a plugin can indeed be compiled as an AAR before they are published.  

Given that the original motivation was to make the AndroidX transition smother

**Comment by hanabi1224** [comment]
Everytime I manually add below line to dependencies section in build.gradle
compileOnly files(System.env.FLUTTER_ROOT + '/bin/cache/artifacts/engine/android-x64/flutter.jar')

**Comment by blasten** [comment]
This will require significant investment from our team. Not just generating the AARs, but also distributing the artifacts on a centralized Maven repo in a somewhat automated way, so we don't put the burden on the plugin author.

**Comment by blasten** [comment]
Refactoring the way Android plugins are included in Flutter won't provide additional benefits at the moment.
The task itself is expensive to implement. It will also increase build times, will require different code paths for local plugin development, and maintaining a service for hosting 3P maven packages.

Due to the above, this feature is out-of-scope, and unlikely to be implemented.
